File: IBlockerKit/Sources/IBlockerKit/Log/StatsStore.swift
import Foundation

public struct DayCounters: Codable, Sendable, Equatable {
    public var total: UInt64
    public var blocked: UInt64

    public init(total: UInt64 = 0, blocked: UInt64 = 0) {
        self.total = total
        self.blocked = blocked
    }
}

/// Cumulative counters, persisted as JSON in the App Group. The tunnel keeps
/// the authoritative copy in memory and flushes on the log-ring cadence; the
/// app reads the file when the tunnel is not reachable.
public struct BlockerStats: Codable, Sendable, Equatable {
    public var totalQueries: UInt64
    public var totalBlocked: UInt64
    public var days: [String: DayCounters]

    public init(totalQueries: UInt64 = 0, totalBlocked: UInt64 = 0, days: [String: DayCounters] = [:]) {
        self.totalQueries = totalQueries
        self.totalBlocked = totalBlocked
        self.days = days
    }

    public mutating func record(blocked: Bool, on day: String) {
        totalQueries += 1
        var counters = days[day] ?? DayCounters()
        counters.total += 1
        if blocked {
            totalBlocked += 1
            counters.blocked += 1
        }
        days[day] = counters
        if days.count > 60 {
            trim(keepDays: 45)
        }
    }

    public mutating func trim(keepDays: Int) {
        let keep = Set(days.keys.sorted().suffix(keepDays))
        days = days.filter { keep.contains($0.key) }
    }

    public func counters(day: String) -> DayCounters {
        days[day] ?? DayCounters()
    }

    /// Local-calendar day key, e.g. "2026-08-09".
    public static func dayKey(for date: Date = Date(), calendar: Calendar = .current) -> String {
        let parts = calendar.dateComponents([.year, .month, .day], from: date)
        return String(format: "%04d-%02d-%02d", parts.year ?? 0, parts.month ?? 0, parts.day ?? 0)
    }
}

public enum StatsPersistence {
    public static func load(from url: URL) -> BlockerStats {
        guard let data = try? Data(contentsOf: url),
              let stats = try? JSONDecoder().decode(BlockerStats.self, from: data) else {
            return BlockerStats()
        }
        return stats
    }

    public static func save(_ stats: BlockerStats, to url: URL) {
        if let data = try? JSONEncoder().encode(stats) {
            try? data.write(to: url, options: .atomic)
        }
    }
}

File: android/core/src/main/kotlin/com/iblocker/core/log/StatsStore.kt
package com.iblocker.core.log

import com.iblocker.core.json.Json
import com.iblocker.core.json.asLong
import com.iblocker.core.json.asObject
import com.iblocker.core.rules.CompiledBlocklist
import java.io.File
import java.time.Instant
import java.time.ZoneId

data class DayCounters(var total: Long = 0, var blocked: Long = 0)

/**
 * Cumulative counters, persisted as JSON. The VPN service keeps the
 * authoritative copy in memory and flushes on the log-ring cadence; the UI
 * reads the file when the service is not running.
 */
class BlockerStats(
    var totalQueries: Long = 0,
    var totalBlocked: Long = 0,
    val days: MutableMap<String, DayCounters> = LinkedHashMap(),
) {
    fun record(blocked: Boolean, day: String) {
        totalQueries += 1
        val counters = days.getOrPut(day) { DayCounters() }
        counters.total += 1
        if (blocked) {
            totalBlocked += 1
            counters.blocked += 1
        }
        if (days.size > 60) trim(keepDays = 45)
    }

    fun trim(keepDays: Int) {
        val keep = days.keys.sorted().takeLast(keepDays).toSet()
        days.keys.retainAll(keep)
    }

    fun counters(day: String): DayCounters = days[day] ?: DayCounters()

    fun toJson(): Map<String, Any?> = mapOf(
        "totalQueries" to totalQueries,
        "totalBlocked" to totalBlocked,
        "days" to days.mapValues { (_, counters) ->
            mapOf("total" to counters.total, "blocked" to counters.blocked)
        },
    )

    companion object {
        /** Local-calendar day key, e.g. "2026-08-09". */
        fun dayKey(epochMillis: Long = System.currentTimeMillis(), zone: ZoneId = ZoneId.systemDefault()): String {
            val date = Instant.ofEpochMilli(epochMillis).atZone(zone).toLocalDate()
            return String.format("%04d-%02d-%02d", date.year, date.monthValue, date.dayOfMonth)
        }

        fun fromJson(value: Any?): BlockerStats {
            val root = value.asObject() ?: return BlockerStats()
            val stats = BlockerStats(
                totalQueries = root["totalQueries"].asLong() ?: 0,
                totalBlocked = root["totalBlocked"].asLong() ?: 0,
            )
            root["days"].asObject()?.forEach { (day, counters) ->
                val entry = counters.asObject() ?: return@forEach
                stats.days[day] = DayCounters(
                    total = entry["total"].asLong() ?: 0,
                    blocked = entry["blocked"].asLong() ?: 0,
                )
            }
            return stats
        }
    }
}

object StatsPersistence {
    fun load(file: File): BlockerStats = try {
        if (file.isFile) BlockerStats.fromJson(Json.parseOrNull(file.readText())) else BlockerStats()
    } catch (_: Exception) {
        BlockerStats()
    }

    fun save(stats: BlockerStats, file: File) {
        runCatching {
            CompiledBlocklist.writeAtomically(file, Json.write(stats.toJson()).toByteArray(Charsets.UTF_8))
        }
    }
}
